[ONNX][Reland] Export node and value with scope name

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #82040

Introduce `_jit_pass_onnx_assign_node_and_value_names` to parse and assign
scoped name for nodes and values in exported onnx graph.
Module layer information is obtained from `ONNXScopeName` captured in `scope`
attribute in nodes. For nodes, the processed onnx node name are stored in
attribute `onnx_name`. For values, the processed onnx output name are stored
as `debugName`.

BowenBao committed:
[ONNX] Export node and value with scope name

[ghstack-poisoned]

justinchuby commented on torch/csrc/onnx/init.cpp:
```diff
@@ -223,6 +223,10 @@ void initONNXBindings(PyObject* module) {
           "Write `args` to the previously specified ONNX log stream.")
+      .def(
+          "_jit_pass_onnx_assign_node_and_value_names",
+          ::torch::jit::onnx::AssignNodeAndValueNames,
```
Use `::torch::wrap_pybind_function` to ensure errors are handled correctly?

BowenBao committed:
Update on "[ONNX] Export node and value with scope name"




[ghstack-poisoned]

titaiwangms reviewed (approved):
:beers:

BowenBao committed:
Update on "[ONNX] Export node and value with scope name"




[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Export node and value with scope name"




[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Export node and value with scope name"




[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Export node and value with scope name"


Introduce `_jit_pass_onnx_assign_node_and_value_names` to parse and assign
scoped name for nodes and values in exported onnx graph.
Module layer information is obtained from `ONNXScopeName` captured in `scope`
attribute in nodes. For nodes, the processed onnx node name are stored in
attribute `onnx_name`. For values, the processed onnx output name are stored
as `debugName`.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Export node and value with scope name"


Introduce `_jit_pass_onnx_assign_node_and_value_names` to parse and assign
scoped name for nodes and values in exported onnx graph.
Module layer information is obtained from `ONNXScopeName` captured in `scope`
attribute in nodes. For nodes, the processed onnx node name are stored in
attribute `onnx_name`. For values, the processed onnx output name are stored
as `debugName`.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Export node and value with scope name"


Introduce `_jit_pass_onnx_assign_node_and_value_names` to parse and assign
scoped name for nodes and values in exported onnx graph.
Module layer information is obtained from `ONNXScopeName` captured in `scope`
attribute in nodes. For nodes, the processed onnx node name are stored in
attribute `onnx_name`. For values, the processed onnx output name are stored
as `debugName`.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Export node and value with scope name"


Introduce `_jit_pass_onnx_assign_node_and_value_names` to parse and assign
scoped name for nodes and values in exported onnx graph.
Module layer information is obtained from `ONNXScopeName` captured in `scope`
attribute in nodes. For nodes, the processed onnx node name are stored in
attribute `onnx_name`. For values, the processed onnx output name are stored
as `debugName`.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Export node and value with scope name"


Introduce `_jit_pass_onnx_assign_node_and_value_names` to parse and assign
scoped name for nodes and values in exported onnx graph.
Module layer information is obtained from `ONNXScopeName` captured in `scope`
attribute in nodes. For nodes, the processed onnx node name are stored in
attribute `onnx_name`. For values, the processed onnx output name are stored
as `debugName`.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Export node and value with scope name"


Introduce `_jit_pass_onnx_assign_node_and_value_names` to parse and assign
scoped name for nodes and values in exported onnx graph.
Module layer information is obtained from `ONNXScopeName` captured in `scope`
attribute in nodes. For nodes, the processed onnx node name are stored in
attribute `onnx_name`. For values, the processed onnx output name are stored
as `debugName`.

[ghstack-poisoned]

justinchuby commented on torch/csrc/jit/passes/onnx/naming.cpp:
```diff
@@ -75,6 +75,131 @@ bool isCompatibleScope(torch::jit::ScopePtr scope) {
+
+std::string NodeNameGenerator::UpdateBaseName(
+    std::unordered_map<std::string, size_t>& base_name_count,
+    std::string base_name) {
```
NOTE: naming

BowenBao commented on torch/csrc/jit/passes/onnx/naming.cpp:
```diff
@@ -75,6 +75,131 @@ bool isCompatibleScope(torch::jit::ScopePtr scope) {
+    auto full_scope_name =
+        ONNXScopeName::variableNameFromRoot(scope, layer_separator_);
+    full_scope_names_[scope] =
+        UpdateBaseName(base_scope_name_counts_, full_scope_name);
```
TODO: change `UpdateBaseName` function name

BowenBao commented on torch/onnx/utils.py:
```diff
@@ -1279,8 +1257,34 @@ def _track_module_attributes_forward_hook(module, input, output):
             m.register_forward_hook(_track_module_attributes_forward_hook)
             m.register_forward_pre_hook(_track_module_attributes_forward_pre_hook)
 
+    def __unqualified_variable_name(qualified_name: str):
```
TODO: move to first PR in stack

justinchuby replied:
```suggestion
    def __unqualified_variable_name(qualified_name: str) -> str:
```

BowenBao commented on torch/csrc/onnx/onnx.h:
```diff
@@ -16,5 +16,7 @@ enum class TrainingMode {
   TRAINING, // Training mode
 };
 
+const std::string OnnxNodeNameAttribute = "onnx_name";
```
TODO: k

BowenBao commented on torch/onnx/utils.py:
```diff
@@ -1279,8 +1257,34 @@ def _track_module_attributes_forward_hook(module, input, output):
+        idx = next(
+            (
+                i
+                for i, atom in reversed(list(enumerate(name_atoms)))
```
```python
for i, atom in reversed(list(enumerate(name_atoms))):
    if not atom.isnumeric():
        return ".".join(name_atoms[i:])
return qualified_name
```

BowenBao committed:
Update on "[ONNX] Export node and value with scope name"


Introduce `_jit_pass_onnx_assign_node_and_value_names` to parse and assign
scoped name for nodes and values in exported onnx graph.
Module layer information is obtained from `ONNXScopeName` captured in `scope`
attribute in nodes. For nodes, the processed onnx node name are stored in
attribute `onnx_name`. For values, the processed onnx output name are stored
as `debugName`.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Export node and value with scope name"


Introduce `_jit_pass_onnx_assign_node_and_value_names` to parse and assign
scoped name for nodes and values in exported onnx graph.
Module layer information is obtained from `ONNXScopeName` captured in `scope`
attribute in nodes. For nodes, the processed onnx node name are stored in
attribute `onnx_name`. For values, the processed onnx output name are stored
as `debugName`.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Export node and value with scope name"


Introduce `_jit_pass_onnx_assign_node_and_value_names` to parse and assign
scoped name for nodes and values in exported onnx graph.
Module layer information is obtained from `ONNXScopeName` captured in `scope`
attribute in nodes. For nodes, the processed onnx node name are stored in
attribute `onnx_name`. For values, the processed onnx output name are stored
as `debugName`.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Export node and value with scope name"


Introduce `_jit_pass_onnx_assign_node_and_value_names` to parse and assign
scoped name for nodes and values in exported onnx graph.
Module layer information is obtained from `ONNXScopeName` captured in `scope`
attribute in nodes. For nodes, the processed onnx node name are stored in
attribute `onnx_name`. For values, the processed onnx output name are stored
as `debugName`.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Export node and value with scope name"


Introduce `_jit_pass_onnx_assign_node_and_value_names` to parse and assign
scoped name for nodes and values in exported onnx graph.
Module layer information is obtained from `ONNXScopeName` captured in `scope`
attribute in nodes. For nodes, the processed onnx node name are stored in
attribute `onnx_name`. For values, the processed onnx output name are stored
as `debugName`.

[ghstack-poisoned]

RalphMao commented:
This new feature will be super helpful! Looking forward to it!

BowenBao committed:
Update on "[ONNX] Export node and value with scope name"


Introduce `_jit_pass_onnx_assign_node_and_value_names` to parse and assign
scoped name for nodes and values in exported onnx graph.
Module layer information is obtained from `ONNXScopeName` captured in `scope`
attribute in nodes. For nodes, the processed onnx node name are stored in
attribute `onnx_name`. For values, the processed onnx output name are stored
as `debugName`.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Export node and value with scope name"


Introduce `_jit_pass_onnx_assign_node_and_value_names` to parse and assign
scoped name for nodes and values in exported onnx graph.
Module layer information is obtained from `ONNXScopeName` captured in `scope`
attribute in nodes. For nodes, the processed onnx node name are stored in
attribute `onnx_name`. For values, the processed onnx output name are stored
as `debugName`.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Export node and value with scope name"


Introduce `_jit_pass_onnx_assign_node_and_value_names` to parse and assign
scoped name for nodes and values in exported onnx graph.
Module layer information is obtained from `ONNXScopeName` captured in `scope`
attribute in nodes. For nodes, the processed onnx node name are stored in
attribute `onnx_name`. For values, the processed onnx output name are stored
as `debugName`.

[ghstack-poisoned]

BowenBao commented:
@abock for approval

BowenBao committed:
Update on "[ONNX] Export node and value with scope name"


Introduce `_jit_pass_onnx_assign_node_and_value_names` to parse and assign
scoped name for nodes and values in exported onnx graph.
Module layer information is obtained from `ONNXScopeName` captured in `scope`
attribute in nodes. For nodes, the processed onnx node name are stored in
attribute `onnx_name`. For values, the processed onnx output name are stored
as `debugName`.

[ghstack-poisoned]

BowenBao commented:
@pytorchbot merge -g

BowenBao committed:
Update on "[ONNX] Export node and value with scope name"


Introduce `_jit_pass_onnx_assign_node_and_value_names` to parse and assign
scoped name for nodes and values in exported onnx graph.
Module layer information is obtained from `ONNXScopeName` captured in `scope`
attribute in nodes. For nodes, the processed onnx node name are stored in
attribute `onnx_name`. For values, the processed onnx output name are stored
as `debugName`.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Export node and value with scope name"


Introduce `_jit_pass_onnx_assign_node_and_value_names` to parse and assign
scoped name for nodes and values in exported onnx graph.
Module layer information is obtained from `ONNXScopeName` captured in `scope`
attribute in nodes. For nodes, the processed onnx node name are stored in
attribute `onnx_name`. For values, the processed onnx output name are stored
as `debugName`.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Export node and value with scope name"


Introduce `_jit_pass_onnx_assign_node_and_value_names` to parse and assign
scoped name for nodes and values in exported onnx graph.
Module layer information is obtained from `ONNXScopeName` captured in `scope`
attribute in nodes. For nodes, the processed onnx node name are stored in
attribute `onnx_name`. For values, the processed onnx output name are stored
as `debugName`.

[ghstack-poisoned]

BowenBao commented:
@pytorchbot merge -g

malfet commented:
This change were reverted internally (though it has not propagated to github yet), due to the global constructor in the header, i.e. this line
https://github.com/pytorch/pytorch/blob/df523a6eeef283a79062f828c3fedce2cc3e32f0/torch/csrc/onnx/onnx.h#L19

Global constructors are bad, but global constructors in headers are especially , will take an AI to adjust our linters to make those a fatal error.

P.S. what's wrong with adding it as `const char*`? Or (though I don't know if C++14 standard allows that) `constexpr std::string`?

weiwangmeta commented:
Manually trigger the revert process trying to resolve sync issues. 
@pytorchbot revert -m="Diff reverted internally" -c="ghfirst"

BowenBao commented:
@malfet thanks for info, switching to `constexpr char []`.

BowenBao committed:
Update on "[ONNX][Reland] Export node and value with scope name"


Introduce `_jit_pass_onnx_assign_node_and_value_names` to parse and assign
scoped name for nodes and values in exported onnx graph.
Module layer information is obtained from `ONNXScopeName` captured in `scope`
attribute in nodes. For nodes, the processed onnx node name are stored in
attribute `onnx_name`. For values, the processed onnx output name are stored
as `debugName`.

[ghstack-poisoned]

BowenBao commented:
@pytorchbot merge -g
